src/sql/src/plan/typeconv.rs

```diff
@@ -985,15 +988,45 @@ pub fn to_jsonb(ecx: &ExprContext, expr: HirScalarExpr) -> HirScalarExpr {
 /// [union-type-conv]: https://www.postgresql.org/docs/12/typeconv-union-case.html
 pub fn guess_best_common_type(
     ecx: &ExprContext,
-    types: &[Option<ScalarType>],
+    types: &[CoercibleScalarType],
 ) -> Result<ScalarType, PlanError> {
     // This function is a translation of `select_common_type` in PostgreSQL with
     // the addition of our near match logic, which supports Materialize
     // non-linear type promotions.
     // https://github.com/postgres/postgres/blob/d1b307eef/src/backend/parser/parse_coerce.c#L1288-L1308
 
+    // If every type is a literal record with the same number of fields, the
+    // best common type is a record with that number of fields. We recursively
+    // guess the best type for each field.
+    if let Some(CoercibleScalarType::Record(field_tys)) = types.first() {
+        if types
+            .iter()
+            .all(|t| matches!(t, CoercibleScalarType::Record(_)))
```

Comment:
This seems like it also wants to check that all of the types are records of the same arity.

Reply:
Ah, sure. The arity gets checked during the actual cast but doesn't hurt to check here too.